### PR TITLE
feat(dataframe): support "__repr__" method in DataFrame

### PR DESCRIPTION
**Pull request by AChenQ**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2b4rt3j 实现 DataFrame.__repr__ 可视化方法](https://app.clickup.com/t/2b4rt3j)

File: graviti/dataframe/frame.py [hunk 1]
```diff
@@ -140,6 +163,38 @@ def _construct(
         obj._index = index
         return obj
 
+    @classmethod
```
Comment (linjiX) [inline graviti/dataframe/frame.py:166]: why this is a `classmethod`?

File: graviti/dataframe/frame.py [hunk 2]
```diff
@@ -140,6 +163,38 @@ def _construct(
...
+    def _load_repr_info(
```
Comment (linjiX) [inline graviti/dataframe/frame.py:167]: Can we make this method constant? (A method will not change the input argument?)

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 3]
```diff
@@ -140,6 +163,38 @@ def _construct(
...
+        data: List[ColumnSeries],
+        *,
+        level: int = 0,
+        start: int = 0,
+    ) -> Tuple[Dict[int, Dict[str, Dict[str, int]]], List[ColumnSeries], int]:
+        items = {}
+        for key, value in values.items():
+            items[key] = {"start": start}
+            if isinstance(value, DataFrame):
+                _, _, start = cls._load_repr_info(
```
Comment (linjiX) [inline graviti/dataframe/frame.py:180]: seems only the last item in the return value is used?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 4]
```diff
@@ -111,7 +104,37 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
...
+                header[index][info["start"] + 1] = f"{name:<{_MAX_REPE_NUMBER}}"
+        repr_lines = ["".join(line) for line in header]
+
+        length = self.__len__()
+        repr_indices: Iterable[int]
+        if self._index._indices is None:
+            if length >= _MAX_REPE_NUMBER:
+                repr_indices = range(_MAX_REPE_NUMBER)
+            else:
+                repr_indices = range(length)
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:123]: `repr_indices = range(min(length, _MAX_REPE_NUMBER))`?

File: graviti/dataframe/frame.py [hunk 5]
```diff
@@ -25,6 +25,9 @@
 from graviti.dataframe.row.series import Series as RowSeries
 from graviti.utility.lazy import LazyFactory
 
+_MAX_REPE_NUMBER = 10
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:28]: `REPE` or `REPR`?
And which one is more accurate, `NUMBER` or `LENGTH` or `ROWS`?

File: graviti/dataframe/frame.py [hunk 6]
```diff
@@ -140,6 +163,38 @@ def _construct(
...
+                )
+            else:
+                data.append(value)
+                items[key]["length"] = 1
+                start += 1
+
+        if level not in result:
+            result[level] = items
+        else:
+            result[level].update(items)
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:195]: Is it possible to use `defaultdict` here?

**Review by Lee-000 — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 7]
```diff
@@ -43,18 +46,9 @@ class DataFrame:
         >>> data = {"col1": [1, 2], "col2": [3, 4]}
         >>> df = DataFrame(data=data)
         >>> df
```
Comment (graczhual) [inline graviti/dataframe/frame.py:48]: add a multi-index example

**Review by graczhual — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 8]
```diff
@@ -140,6 +195,29 @@ def _construct(
...
+        if self._index._indices is None:
+            return range(min(length, _MAX_REPE_NUMBER))
+
+        if length >= _MAX_REPR_ROWS:
+            return self._index._indices[:_MAX_REPR_ROWS]
+
+        return self._index._indices
+
+    @classmethod
+    def from_lazy_factory(cls: Type[_T], factory: LazyFactory) -> _T:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:210]: what is this method for?

File: graviti/dataframe/frame.py [hunk 9]
```diff
@@ -25,6 +26,10 @@
 from graviti.dataframe.row.series import Series as RowSeries
 from graviti.utility.lazy import LazyFactory
 
+_MAX_REPE_NUMBER = 3
```
Comment (linjiX) [inline graviti/dataframe/frame.py:29]: REPR？

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 10]
```diff
@@ -104,29 +112,76 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Union[ColumnSeries, "Da
...
-
-        Return:
-            The loaded :class:`~graviti.dataframe.DataFrame` object.
+    @staticmethod
+    def _get_repr_header(
+        values: Dict[str, Union["DataFrame", ColumnSeries]],
+        *,
+        level: int = 0,
+        start: int = 0,
+    ) -> Tuple[Dict[int, Dict[str, Dict[str, int]]], List[ColumnSeries], int]:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:164]: the return type is too complicate, can we simplify it? or maybe add some doc?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 11]
```diff
@@ -6,6 +6,7 @@
 """The implementation of the Graviti DataFrame."""
 
 
+from collections import defaultdict
```
Comment (linjiX) [inline graviti/dataframe/frame.py:9]: `function` -> `method` in commit message.
normally method means the function in class. function means standalone function

File: graviti/dataframe/frame.py [hunk 12]
```diff
@@ -104,29 +112,76 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Union[ColumnSeries, "Da
         if isinstance(key, str):
             return self._columns[key]
 
-        new_columns = {name: self._columns[name] for name in key}
-        return self._construct(new_columns, self._index)
+        if isinstance(key, Iterable):
```
Comment (linjiX) [inline graviti/dataframe/frame.py:115]: the code should be rebased.

File: graviti/dataframe/frame.py [hunk 13]
```diff
@@ -104,29 +112,76 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Union[ColumnSeries, "Da
...
-    @classmethod
-    def from_lazy_factory(cls: Type[_T], factory: LazyFactory) -> _T:
-        """Create DataFrame with lazy factory and schema.
-
-        Arguments:
-            factory: class :class:`~graviti.utility.lazy.LazyFactory` instance.
-
-        Return:
-            The loaded :class:`~graviti.dataframe.DataFrame` object.
+    @staticmethod
```
Comment (linjiX) [inline graviti/dataframe/frame.py:158]: why `staticmethod`? why not use `self` to get the `values`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 14]
```diff
@@ -140,6 +195,29 @@ def _construct(
...
         return obj
 
+    def _get_repr_indices(self) -> Iterable[int]:
+        length = self.__len__()
+        # pylint: disable=protected-access
+        if self._index._indices is None:
+            return range(min(length, _MAX_REPE_NUMBER))
+
+        if length >= _MAX_REPR_ROWS:
+            return self._index._indices[:_MAX_REPR_ROWS]
```
Comment (linjiX) [inline graviti/dataframe/frame.py:205]: Is `islice` better?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 9 above)
Comment (linjiX) [reply graviti/dataframe/frame.py:29]: What is the difference between REPR_NUMBER and REPR_ROWS?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 15]
```diff
@@ -104,18 +112,90 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Union[ColumnSeries, "Da
...
+
+        raise KeyError(key)
 
     def __setitem__(self, key: str, value: Iterable[Any]) -> None:
         pass
 
     def __repr__(self) -> str:
-        return f"\n DataFrame\n columns: {self._columns} \n index: {self._index}"
+        header_info, data, _ = self._get_repr_header(self._columns)
+        repr_length = [_DEFAULT_REPR_NUMBER] * (len(data) + 1)
```
Comment (linjiX) [inline graviti/dataframe/frame.py:126]: why length is a list?

File: graviti/dataframe/frame.py [hunk 16]
```diff
@@ -104,18 +112,90 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Union[ColumnSeries, "Da
...
+            [_PLACEHOLDER] * (len(data) + 1) for _ in range(len(header_info))
+        ]
+        for index, value in header_info.items():
+            for name, info in value.items():
+                repr_table[index][info["start"] + 1] = name
+                repr_length[info["start"] + 1] = max(len(name), repr_length[info["start"] + 1])
+
+        for index in self._get_repr_indices():
+            cur_line = [index]
+            for idx, column in enumerate(data):
```
Comment (linjiX) [inline graviti/dataframe/frame.py:138]: why we have `idx` and `index` at the same place?

File: graviti/dataframe/frame.py [hunk 17]
```diff
@@ -104,18 +112,90 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Union[ColumnSeries, "Da
...
+        repr_table: List[Union[List[int], List[str]]] = [
+            [_PLACEHOLDER] * (len(data) + 1) for _ in range(len(header_info))
+        ]
+        for index, value in header_info.items():
+            for name, info in value.items():
+                repr_table[index][info["start"] + 1] = name
+                repr_length[info["start"] + 1] = max(len(name), repr_length[info["start"] + 1])
+
+        for index in self._get_repr_indices():
+            cur_line = [index]
```
Comment (linjiX) [inline graviti/dataframe/frame.py:137]: what is `cur` meaning?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 18]
```diff
@@ -113,11 +122,80 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
...
+
+        repr_lines = [
+            "".join(f"{item:<{repr_length[index]+1}}" for index, item in enumerate(line))
+            for line in repr_table
+        ]
+        if self.__len__() > _DEFAULT_REPR_NUMBER:
+            repr_lines.append(f"...({self.__len__()})")
+        return "\n".join(repr_lines)
+
+    def __format__(self, format_spec: str) -> str:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:151]: why using `__format__`???

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 19]
```diff
@@ -113,11 +123,77 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
...
+        string = f"DataFrame{self.shape}"
```
Comment (linjiX) [inline graviti/dataframe/frame.py:153]: Is DataFrame.shape not implemented.

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 18 above)
Comment (AChenQ) [reply graviti/dataframe/frame.py:156]: repr use `__format__` to display nested Dataframe.

File: graviti/dataframe/frame.py [hunk 20]
```diff
@@ -200,7 +291,7 @@ def loc(self) -> DataFrameLocIndexer:
     def shape(self) -> Tuple[int, int]:
         """Return a tuple representing the dimensionality of the DataFrame.
 
-        Return:
+        Returns:
             Shape of the DataFrame.
 
         Examples:
```
Comment (graczhual) [inline graviti/dataframe/frame.py:297]: add a multi-index example? This is one difference between us and Pandas

File: graviti/dataframe/frame.py [hunk 21]
```diff
@@ -113,11 +123,81 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
...
+        for key, value in self._columns.items():
+            items[key] = {"start": start}
+            if isinstance(value, DataFrame):
+                # pylint: disable=protected-access
+                sub_result, sub_data, start = value._get_repr_header(
+                    level=level + 1,
+                    start=start,
+                )
+                header_info.update(sub_result)
+                data += sub_data
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:192]: Which one is better, `extend` or `+=`?

File: graviti/dataframe/frame.py (hunk 18 above)
Comment (linjiX) [reply graviti/dataframe/frame.py:156]: will it breaks the normal `format` method?

File: graviti/dataframe/frame.py [hunk 22]
```diff
@@ -113,11 +123,81 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
...
+        repr_length = [_DEFAULT_REPR_LETTER] * (len(data) + 1)
+
+        repr_table: List[List[Union[int, str]]] = [
+            [_PLACEHOLDER] * (len(data) + 1) for _ in range(len(header_info))
+        ]
+        # load header to repr_table
+        for index, value in header_info.items():
+            for name, info in value.items():
+                repr_table[index][info["start"] + 1] = name
+                repr_length[info["start"] + 1] = max(len(name), repr_length[info["start"] + 1])
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:136]: Duplicate calculation for `info["start"] + 1`?

**Review by Lee-000 — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 23]
```diff
@@ -113,7 +123,38 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
...
+        repr_length = [_DEFAULT_REPR_LETTER] * (len(data) + 1)
+
+        repr_table: List[List[Union[int, str]]] = [
+            [_PLACEHOLDER] * (len(data) + 1) for _ in range(len(header_info))
+        ]
+        # load header to repr_table
+        for index, value in header_info.items():
+            for name, info in value.items():
+                repr_table[index][info["start"]] = name
+                repr_length[info["start"]] = max(len(name), repr_length[info["start"]])
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:136]: Repeatedly get the same value from the dict `info`.

File: graviti/dataframe/frame.py [hunk 24]
```diff
@@ -113,7 +123,38 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
...
+                )
+                repr_length[location] = max(len(item), repr_length[location])
+                data_line.append(item)
+            repr_table.append(data_line)
+
+        repr_lines = [
+            "".join(f"{item:<{repr_length[index]+1}}" for index, item in enumerate(line))
+            for line in repr_table
+        ]
+        if self.__len__() > _DEFAULT_REPR_LETTER:
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:155]: `_DEFAULT_REPR_LETTER` or `_MAX_REPR_ROWS`?

File: graviti/dataframe/frame.py [hunk 25]
```diff
@@ -142,6 +183,58 @@ def _construct(
...
+        return header_info, data, location
+
+    def _repr_folding(self) -> str:
+        return f"DataFrame{self.shape}"
+
+    def _get_repr_indices(self) -> Iterable[int]:
+        length = self.__len__()
+        # pylint: disable=protected-access
+        if self._index._indices is None:
+            return range(min(length, _DEFAULT_REPR_LETTER))
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:231]: `_DEFAULT_REPR_LETTER` or `_MAX_REPR_ROWS`?

**Review by Lee-000 — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 26]
```diff
@@ -200,15 +293,27 @@ def loc(self) -> DataFrameLocIndexer:
...
+                )
+            >>> df
+                filename box2ds
+                         x      y
+            0   a.jpg    1      1
+            1   b.jpg    2      2
+            2   c.jpg    3      3
             >>> df.shape
-            (2, 2)
+            (2, 3)
```
Comment (graczhual) [inline graviti/dataframe/frame.py:313]: (3,2)?

**Review by graczhual — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 26 above)
Comment (AChenQ) [reply graviti/dataframe/frame.py:313]:
```
pd.DataFrame([[1, 2, 3], [3, 4, 5]]).shape
(2, 3)
```

**Review by graczhual — APPROVED** [review]

**Review by Lee-000 — APPROVED** [review]

File: graviti/dataframe/frame.py (hunk 15 above)
Comment (linjiX) [reply graviti/dataframe/frame.py]: normally length should be a `int`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 27]
```diff
@@ -142,6 +184,57 @@ def _construct(
...
+
+        """
+        header_info: Dict[int, Dict[str, Dict[str, int]]] = defaultdict(dict)
+        data: List[ColumnSeries] = []
+        items = {}
+        for key, value in self._columns.items():
+            items[key] = {"location": location}
+            if isinstance(value, DataFrame):
+                # pylint: disable=protected-access
+                sub_result, sub_data, location = value._get_repr_information(
```
Comment (linjiX) [inline graviti/dataframe/frame.py:211]: `sub_result`? why not using `sub_header_info`?

**Review by linjiX — CHANGES_REQUESTED** [review]

**Comment by linjiX** [comment]
Try to reconsider the structure of `header_info`,
And try to find a better variable name for every variable in the code.

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 28]
```diff
@@ -142,6 +183,43 @@ def _construct(
...
 
+    def _flatten(self) -> Tuple[List[Tuple[str]], List[ColumnSeries]]:
+        """Flatten the nested DataFrame.
+
+        Returns:
+            header_info: The flattened header.
+            data: The flattened data.
+
+        """
+        flatten_header: List[Tuple[str]] = []
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:194]: Only one item in the tuple?

File: graviti/dataframe/frame.py [hunk 29]
```diff
@@ -142,6 +183,43 @@ def _construct(
         obj._index = index
         return obj
 
+    def _flatten(self) -> Tuple[List[Tuple[str]], List[ColumnSeries]]:
+        """Flatten the nested DataFrame.
+
+        Returns:
+            header_info: The flattened header.
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:190]: `header_info` or `header`?

File: graviti/dataframe/frame.py [hunk 30]
```diff
@@ -113,7 +121,40 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
         pass
 
     def __repr__(self) -> str:
-        return f"\n DataFrame\n columns: {self._columns} \n index: {self._index}"
+        flatten_header, flatten_data = self._flatten()
+        header_depth = max(len(column_header) for column_header in flatten_header)
+        column_length = len(flatten_data) + 1
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:126]: Is `length` accurate? Would `number` be better?

**Review by Lee-000 — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 31]
```diff
@@ -142,6 +183,43 @@ def _construct(
...
+        Returns:
+            header_info: The flattened header.
+            data: The flattened data.
+
+        """
+        flatten_header: List[Tuple[str]] = []
+        flatten_data: List[ColumnSeries] = []
+        for key, value in self._columns.items():
+            if isinstance(value, DataFrame):
+                sub_header, sub_data = value._flatten()  # pylint: disable=protected-access
```
Comment (linjiX) [inline graviti/dataframe/frame.py:198]: will `nested_header` better than `sub_header`

File: graviti/dataframe/frame.py [hunk 32]
```diff
@@ -113,7 +121,40 @@ def __setitem__(self, key: str, value: Iterable[Any]) -> None:
         pass
 
     def __repr__(self) -> str:
-        return f"\n DataFrame\n columns: {self._columns} \n index: {self._index}"
+        flatten_header, flatten_data = self._flatten()
+        header_depth = max(len(column_header) for column_header in flatten_header)
```
Comment (linjiX) [inline graviti/dataframe/frame.py:125]: Is `map` better?
```python
max(map(len, flatten_header))
```

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 28 above)
Comment (AChenQ) [reply graviti/dataframe/frame.py:194]: There is only str in this tuple...

**Review by Lee-000 — APPROVED** [review]

File: graviti/dataframe/frame.py [hunk 33]
```diff
@@ -142,6 +176,49 @@ def _construct(
...
+                header.extend((key, *sub_column) for sub_column in nested_header)
+                data.extend(nested_data)
+            else:
+                data.append(value)
+                header.append((key,))
+
+        return header, data
+
+    def _repr_folding(self) -> str:
+        return f"DataFrame{self.shape}"
```
Comment (linjiX) [inline graviti/dataframe/frame.py:194]: use `self.__class__.__name__` to replace `DataFrame`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 34]
```diff
@@ -142,6 +176,49 @@ def _construct(
...
+        if length >= _MAX_REPR_ROWS:
+            return islice(self._index._indices, _MAX_REPR_ROWS)
+
+        return self._index._indices
+
+    def _get_repr_body(self, flatten_data: List[ColumnSeries]) -> List[List[str]]:
+        lines = []
+        for i in self._get_repr_indices():
+            line: List[str] = [str(i)]
+            for column in flatten_data:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:211]: can we use `zip(*flatten_data)`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 34 above)
Comment (AChenQ) [reply graviti/dataframe/frame.py:208]: Series not support `__iter__` yet.

**Review by linjiX — APPROVED** [review]

**Pull request merged**